### DISCUSSION TITLE
Errors Navigation from PMLC Logs

### DISCUSSION THREAD

**Discussion by tajmone** [pml-lang/pml-companion#35] [Editors Support]
Most modern general purpose code editors and IDEs allow parsing error logs of build operations to map errors to their locations in the source files, thus enabling users to see both the error log and click on each error to jump its location in the source files, as well as offering some pop-up tooltips with the extracted error info in the source files.

This indeed a powerful and desirable feature to add to any PML syntax for third party editors.

For example, Sublime Text exposes this feature in its [Build Systems] via the the [`line_regex`][line_regex] option.

(See: tajmone/Sublime-PML#16)

# Sublime Alan Example

You can see how I've implemented this feature in [Sublime Alan], a custom syntax for the ALAN IF language:

![screenshot 1](https://raw.githubusercontent.com/tajmone/sublime-alan/master/screenshots/Build_Errors_Editor.png)

As you can see from the following screenshot, the error report from the compiler is also syntax highlighted to simplify visual inspection of the errors, warning or info messages.

![screenshot 2](https://raw.githubusercontent.com/tajmone/sublime-alan/master/screenshots/Build_Errors_Console.png)

Each error in the log can be clicked to open the culprit source file at the indicate line and column, and the errors are also shown in the source files via pop-up carrying error info.

All errors can be quickly navigated through using default keyboard shortcuts to move to next or previous error in the list.

Similar results should be achievable in most modern editors (e.g. VSCode, Atom, etc.), not just Sublime Text.

# Feature Requirement

In order to correctly implement this feature, the error/info log produced by PMLC needs to adopt a consistent messaging format, because the feature uses RegEx patterns to scan the report message, line by line.

This means that editor syntaxes developer need PMLC to offer guarantees regarding its adopted messaging format, and that it will not change over time, and that this format is also easy to parse via RegExs in order to capture the required fields, which are:

1. filename
1. line number
1. column number
1. message

This is something that should probably be documented somewhere (either in the _PML Reference Manual_ or in the Wiki of the PMLC repository), so that whoever is working on a new PML syntax can be guided on the required RegEx(s) and be assured regarding the consistency of errors formatting in the course of time.

# General Considerations

I've noticed that currently PMLC will abort at the first error encountered. Is this a design choice or a temporary state of affairs? Is it because the first error could jeopardize correct parsing of the rest of the document, leading to incorrect error messages that wouldn't be there if the first error was fixed?

Obviously, being informed of all the errors in a single PMLC invocation would be preferable, for it would avoid having to re-attempt the conversion after each error fix. But if the price of this would be the risk of receiving false positive errors than it might not be worth it.

An important question is whether PMLC is going to report just **errors** are also **warnings** and, maybe, **info** messages in the future. I.e. if some options will be added to enable warnings or extra info it would be worth capturing them from PMLC report, especially if these can be mapped to specific position in a source file (although some **info** message might not refer to any specific position, but just pertain to statistical data regarding the document).

This Discussions thread could explore further the topic of what could be expected from PMLC reports, and which error/info formatting convention might be better for capturing errors info in PML syntaxes for various editors.

-------------------------------------------------------------------------------

### References

- [ST3 Docs]:
    + [Build Systems]:
        * [`file_regex`][file_regex] option.
        * [`line_regex`][line_regex] option.
- [ST3 Community Docs]:
    + [Guide](https://docs.sublimetext.io/guide/):
        * [Build Systems](https://docs.sublimetext.io/guide/usage/build-systems.html)

[ST3 Docs]: https://www.sublimetext.com/docs/3/
[Build Systems]: https://www.sublimetext.com/docs/3/build_systems.html
[file_regex]: https://www.sublimetext.com/docs/3/build_systems.html#exec_option-file_regex
[line_regex]: https://www.sublimetext.com/docs/3/build_systems.html#exec_option-line_regex

[ST3 Community Docs]: https://docs.sublimetext.io

[Sublime Alan]: https://github.com/tajmone/sublime-alan#build-systems

**Comment by pml-lang** [comment]
This is an important topic. Good error handling improves the user experience, and increases productivity. Thanks for starting this discussion.

# Current Error Handling

The current version (1.4.0.) of the PML converter behaves like this:
- The converter aborts at the first error encountered.
- The first error encountered is written to OS err.
- There is no distinction between errors and warnings.

Reasons for such a simplistic error handling:
- easy to implement in an initial proof-of-concept version
- no false positives (i.e. no incorrect errors reported due to a previous correct error)

# Future Error Handling

## Types of errors

There will be three kinds of errors:

- *Warning*
A minor problem that does not prevent the converter to produce output (such as an HTML document).
Example: An empty element in a list (currently an error)

- *Error*
- - *Non-canceling error*
A major problem that does not prevent the parser to create the AST (and detect additional errors), but prevents the converter to produce correct output.
Examples:
an attribute is defined twice (-> the parser simply ignores the second assignment)
an attribute value is illegal (e.g. `border = foo`)
- - *Canceling error*
A major problem that prevents the parser to create the AST. The whole process is immediately cancelled (to avoid false positives).
Examples:
Unclosed comment
Unbalanced bracket pairs
File read error

## Error Reporting

### Default Behavior

By default the converter writes all errors and warnings to OS err, in a user-friendly format (can be redirected to a file with standard OS redirect operator).

### Disable Warnings

A config parameter (e.g. command line option) allows the user to disable all warnings.

### Open Editor For First Error Encountered

A config parameter defines a template for an OS command to open a text editor. The editor will display the file containing the first error encountered (or warning if there are no errors), and the cursor will be positioned at the line and column of the error.

For example, VSCode has the `goto` [command line option](https://code.visualstudio.com/docs/editor/command-line). The template to assign to the converter's command line option would be: `code --goto {{file}}:{{line}}:{{column}}`

I've implemented this in PPL and it works well for all editors that provide this option, without the need for any editor plugins.

### Errors/Warnings Displayed in a GUI

A better solution is to display all errors/warnings in a GUI, with the ability to click on any error/warning to open the editor at the location of the error.

### Log File

The best solution is to produce a log file that can be explored by the editor, so that the editor's native error handling features can be used (highlighting, easy error navigation), like what you did in Sublime Alan.

This is easy to implement in PML, but requires an editor plugin that can read PML's error log file.

IMO the log file should be a CSV file with a format that makes it easy to extract data with regexes (as described in your comment). I'm open to any suggestions for the format. If there is any 'standard' (or rather common format) we should stick to that.

## Note

The new parser I'm currently working on distinguishes already between warnings, non-canceling errors, and canceling errors. For each error/warning encountered, it calls a dedicated function in an error handler that is provided to the parser. The error handler is defined as a Java interface, so that customized error handlers can be programmatically provided to the parser. Hence everything is already well prepared to implement the error handling features described above.

# Statistics

One can easily imagine all kinds of useful statistics generated for authors (e.g. number of words, sentences, paragraphs, chapters; longest sentence, list of words sorted by length, etc. etc.).

While the converter could emit some basic statistics, I think that this feature should rather be implemented by specific plugins or tools. As the new parser (used in version 2.0) is based on pXML, it will be easy to write out the parsed AST in strict pXML or XML format (e.g. converter option `--format pxml` or `--format xml`).  Hence, all existing XML technologies can then be used to explore the AST, as explained in chapter [Using XML technology with PML](https://pxml-lang.github.io/docs/articles/open-source-parser-for-practical-xml/index.html#ch__11) of my article [Open-Source Parser For Practical XML (pXML)](https://pxml-lang.github.io/docs/articles/open-source-parser-for-practical-xml/index.html). Users will be able to use their preferred XML tools and programming languages to query XML documents and create their customized statistics and reports. And they can share and open-source their solutions.

**Reply by tajmone** [reply]
## Current Error Handling

> - There is no distinction between errors and warnings.

So they both output to STDERR and produce non-zero exit code?

> Reasons for such a simplistic error handling:
>
> - easy to implement in an initial proof-of-concept version
> - no false positives (i.e. no incorrect errors reported due to a previous correct error)

Sounds fair enough.

## Future Error Handling

### Types of errors

> - _Warning_
> - _Non-canceling error_

It's not clear whether these will produce a non-zero exit code.

How should an invoker consider these operations, as success or failure?

### Error Reporting

> A config parameter defines a template for an OS command to open a text editor. The editor will display the file containing the first error encountered (or warning if there are no errors), and the cursor will be positioned at the line and column of the error.

Sounds an excellent idea, and any editor/IDE worth its salt should natively support this.

> The best solution is to produce a log file that can be explored by the editor, so that the editor's native error handling features can be used (highlighting, easy error navigation), like what you did in Sublime Alan.
>
> This is easy to implement in PML, but requires an editor plugin that can read PML's error log file.

I'm not sure about how other editors handle this, but in Sublime Text there's no need for a dedicated plug-in, it's supported natively.

Also, for ST it doesn't really make any difference if the error report is captured via I/O STREAMS or read from a log file — different Build options, same approach: just need to define RegEx groups that capture the expected error info (as described above). Ideally, all the error info should be found on the same line (even though extra info maybe shown on the following lines).

Reading error logs from file is supported just to support cases where there's a toolchain in place that needs to store logs on disk; but unless strictly needed, using I/O STREAMS is preferable because it's also much faster — i.e. unless the logs are going to be really long, which probably is not the case with PML, especially if it aborts on first encountered error.

> IMO the log file should be a CSV file with a format that makes it easy to extract data with regexes (as described in your comment). I'm open to any suggestions for the format. If there is any 'standard' (or rather common format) we should stick to that.

I personally don't think we need to sacrifice human readability for this, as long as the error info maps to some symbols which are easy to parse it should be fine. E.g. a formatting convention like this one (the one used by Alan):

    "<filename>.pml", line <ln>(<cn>): <err numb> <err type> : <error desc>

which in a real error could look like:

    "my-cv.pml", line 3(10): 100 W : Malformed date.
    "my-cv.pml", line 20(5): 49 E : File read error.

etc., where each error and warning could have its own number assigned, and a single letter to distinguish between _warning_ (`W`), and _errors_ (`E`) — as for the distinction between cancelling and non-cancelling errors, you could use either an extra letter, or assign different error-number ranges to each type.

The Alan compiler, for example, offers a special option to format error reports in an editor-friendly format, whereas the default format is more human friendly but harder to machine-parse. PML could also adopt a similar approach, if you think it would help.

I'm not quite convinced about CVS making this task easier (as opposed to adopting a strict line-formatting convention, as shown above).

If you look at the [Sublime Alan settings] to capture the compiler errors, they are quite simple and straight forward:

```jsonc
  "file_regex":  "^\"([^\"]+)\", line (\\d+)\\((\\d+)\\): (.*)$",
  //
  // The captured RegEx groups should be:
  //   1. filename
  //   2. line number
  //   3. column number
  //   4. message
```

It's just a one-line RegEx in a JSON-like ([Jsonc]) settings file, nothing more.

The philosophy of most modern general purpose editors, regarding this feature, seems to rely on the tool producing a well-formatted RegEx-friendly error log — and ignore all other cases, for which a dedicated custom plug-in will be needed. This is an example of when editors and IDEs trends tend to influence languages and tools design (e.g. VSCode's LSP having changed the way languages compilers are designed).

So, unless PML needs to generate very complex error messages, I would advise sticking to a simple solution like the one suggested above.

### Statistics

> One can easily imagine all kinds of useful statistics generated for authors (e.g. number of words, sentences, paragraphs, chapters; longest sentence, list of words sorted by length, etc. etc.).

Indeed, this is a wide topic where expectations might largely depend on the type of document being handled, and the context in which its author(s) operate.

Still, there are probably some statistics relating to the supported format(s) that are worth supporting natively — e.g. extracting meta-data, a document description/abstract, query the list of authors, etc.

I.e. some basic info extraction operations might qualify for native support since they could ensure consistently formatted meta-data and other editorial information that might require automation in various contexts (e.g. to automate publishing an article/book on a website which requires to also fill forms with the author(s) name, copyright info, ISBN, a brief description, etc.).

In this case, it would be more reliable to have PML guarantee access to such info, because third party packages might become stale, or introduce proliferation of formats. and a company or publisher might want to rely on a well supported tool to handle similar automation tasks via standardized formats.

> While the converter could emit some basic statistics, I think that this feature should rather be implemented by specific plugins or tools. As the new parser (used in version 2.0) is based on pXML, it will be easy to write out the parsed AST in strict pXML or XML format [...]

Extensibility via third party plug-ins/tools is always preferable to bloating the app with unnecessary features.



[Sublime Alan settings]: https://github.com/tajmone/sublime-alan/blob/master/Alan.sublime-build
[Jsonc]: https://komkom.github.io/

**Reply by pml-lang** [reply]
> So they both output to STDERR and produce non-zero exit code?

Yes, and the exit code is always 1 (current version).

> It's not clear whether these will produce a non-zero exit code.

The exit code will depend on the type of errors encountered:
- no errors and no warnings: 0
- warning(s) but no errors: 0
- non-canceling error(s): 1
- canceling error: 2
- unrecoverable error (e.g. file read error, out of memory, etc.): 3

> a special option to format error reports in an editor-friendly format, whereas the default format is more human friendly but harder to machine-parse.

Yes, PML should do the same, because I can't imagine a single format that is well suited for humans and machines. Also, messages for humans should include the line of PML code where the error occurred, with a marker to show the error location (as done in the current version).

> some basic info extraction operations might qualify for native support since they could ensure consistently formatted meta-data and other editorial information that might require automation in various contexts

True.

**Reply by tajmone** [reply]
> Yes, PML should do the same, because I can't imagine a single format that is well suited for humans and machines. Also, messages for humans should include the line of PML code where the error occurred, with a marker to show the error location (as done in the current version).

As long at the error info line matches the RegEx, and the human-readable description doesn't (this should be easy via the `^` anchor in the RegEx, and use of indentation in the human explanation) the editor's RegEx should be able to ignore all non-error info lines.

In any case, we can quickly set-up Sublime PML to test this when a new PML release is out, or even test it via Sublime Text using a dev branch of the Sublime PML repo and a pre-release PML package (locally).

Even if we don't get this right straight away it shouldn't be a huge deal — it might start as an experimental feature, until proven to be usable on most editors.

In the worst case scenario, you'll have to add a dedicated option to output errors for editors' handling (a bit like Git's porcelain vs plumbing commands); but I'm confident that the solution is going to be actually easier than it looks.
